pandas/core/groupby/grouper.py

```diff
@@ -338,6 +338,20 @@ def _set_grouper(self, obj: FrameOrSeries, sort: bool = False):
             if getattr(self.grouper, "name", None) == key and isinstance(
                 obj, ABCSeries
             ):
+                # In some cases the self._grouper may be sorted differently than obj.
+                # self.obj has the right order with the old index in the first go
+                # around. We align the index from obj with the self.obj index to
+                # select the correct values. Additionally we have to sort
+                # obj.index correctly to aggregate correctly.
+                if not hasattr(self, "_index_mapping"):
+                    self._index_mapping = DataFrame(
+                        self.obj.index, columns=["index"]
+                    ).sort_values(by="index")
```

Comment:
umm, it seems like you can just 
```ax = self._grouper.take(obj.index.sort_values())```?

Reply:
Oh, thanks. Did not remember the codes keyword.

But that does not work here. `self._grouper` is a Index containing all elements from the previous DataFrame, while obj is only a single group from this DataFrame. With codes we would always take the elements 0-len(obj).

We have to keep the whole sorted DataFrame and determine the spots, which are from our current group.

Reply:
use self._grouper.indices (or you may need to get the group_info)

you don't want to recreate things which are already computed

Reply:
This information are not available in resample cases, because self._grouper is an DatetimeIndex in this case. We need the spots of the obj elements in the resorted df, which is sorted the way the DatetimeIndex is.

We could restrict this to the case, when self.grouper is an Index?

Reply:
In the example I am working with, we have the following original df:

Information available to use here:
self.obj in the first go around:
````
      cat  index       date
x1  cat_1     20 2018-02-03
x5  cat_1     30 2018-12-04
x0  cat_1      5 2019-02-01
x3  cat_1      3 2019-02-02
x4  cat_2      4 2019-02-02
x2  cat_2     22 2020-03-11
x6  cat_2     10 2020-03-11
x7  cat_1     50 2020-12-12
````

self._grouper:
````
DatetimeIndex(['2018-02-03', '2018-12-04', '2019-02-01', '2019-02-02',
               '2019-02-02', '2020-03-11', '2020-03-11', '2020-12-12'],
              dtype='datetime64[ns]', name='date', freq=None)
````

and obj for the first group:

````
x0     5
x1    20
x3     3
x5    30
x7    50
````
We have to determine which element from the obj Series correspond with the spots in the DatetimeIndex and we have to reorder the obj Series to align with the DatetimeIndex

Reply:
there a number of ways to do this
try using .get_indexer

we cannot have state like you are doing - not to mention do not use higher level functions inside the groupby

check out core/sorting for appropriate things